tools/dds/py-dds-server/python.cpp

```diff
@@ -209,9 +215,67 @@ PYBIND11_MODULE(NAME, m) {
+    // These are used alongside the 'dds_topic' class:
+    //      topic = dds.topic( p, dds.types.device_info(), "realsense/device-info" )
+    auto types = m.def_submodule( "types", "all the types that " SNAME " can work with" );
+
+    // We need to declare a basic TypeSupport or else dds_topic ctor won't work
+    using eprosima::fastdds::dds::TypeSupport;
+    py::class_< TypeSupport > topic_type( types, "topic_type" );
+
+    // Another submodule, 'raw' is used with dds_topic to actually send and receive messages, e.g.:
+    //      reader = dds.topic_reader( topic )
+    //      ...
+    //      info = dds.device_info( dds.raw.device_info( reader ) )
+    //auto raw = m.def_submodule( "raw", "all the topics that " SNAME " can work with" );
+
+
+    using librealsense::dds::dds_topic;
+    py::class_< dds_topic, std::shared_ptr< dds_topic > >( m, "topic" )
+        .def( py::init< std::shared_ptr< dds_participant > const &,
+                        eprosima::fastdds::dds::TypeSupport const &,
+                        char const * >() )
+        .def( "get_participant", &dds_topic::get_participant )
+        .def( "__repr__", []( dds_topic const & self ) {
+            std::ostringstream os;
+            os << "<" SNAME ".topic \"" << self->get_name() << "\"";
+            os << ">";
+            return os.str();
+        } );
+
+    using librealsense::dds::dds_topic_reader;
+    py::class_< dds_topic_reader, std::shared_ptr< dds_topic_reader > >( m, "topic_reader" )
+        .def( py::init< std::shared_ptr< dds_topic > const & >() )
+        .def( FN_FWD( dds_topic_reader, on_data_available, ( ), ( ), callback(); ) )
+        .def( FN_FWD( dds_topic_reader,
+                      on_subscription_matched,
+                      (),
+                      (eprosima::fastdds::dds::SubscriptionMatchedStatus const &),
+                      callback(); ) )
+        .def( "run", &dds_topic_reader::run );
+
+
+    // The actual types are declared as functions and not classes: the py::init<> inheritance rules are pretty strict
+    // and, coupled with shared_ptr usage, are very hard to get around. This is much simpler...
     using librealsense::dds::topics::device_info;
+    using raw_device_info = librealsense::dds::topics::raw::device_info;
+    types.def( "device_info", []() { return TypeSupport( new device_info::type ); } );
+
     py::class_< device_info >( m, "device_info" )
         .def( py::init<>() )
+        //.def( py::init< raw_device_info const & >() )
+        .def( py::init( []( dds_topic_reader const & reader ) {
+            auto actual_type = reader.topic()->get()->get_type_name();
+            if( actual_type != "librealsense::dds::topics::raw::device_info" )
+                throw std::runtime_error( "can't initialize raw::device_info from " + actual_type );
+            raw_device_info raw_data;
+            eprosima::fastdds::dds::SampleInfo info;
+            DDS_API_CALL( reader->take_next_sample( &raw_data, &info ) );
```

Comment:
Can you explain? the constructor reads messages?

Reply:
Yeah for Python only -- rather than having to deal with raw_device_info and all that crap, and it enforces type-checking.

Reply:
```
def on_data_available():
    global r
    data = dds.device_info( r )
    print( data )

t = dds.topic( p, dds.types.device_info(), "realsense/device-info" )
r = dds.topic_reader( t )
r.on_data_available( on_data_available )
r.run()
```